File: 2015/9/rust/src/main.rs
use std::collections::HashMap;
use std::fs;
use itertools::Itertools;

#[derive(Debug)]
struct Node {
    connections: HashMap<String, i32>
}

impl Node {
    fn new() -> Self {
        Node {
            connections: HashMap::new(),
        }
    }

    fn add_neighbor(&mut self, destination: String, cost: i32){
        self.connections.insert(destination, cost);
    }

    fn get_distance(&self, destination: &str) -> Option<&i32> {
        self.connections.get(destination)
    }
}

fn parse_input(path: &str) -> HashMap<String, Node>{
    let mut nodes = HashMap::new();
    let data = fs::read_to_string(path).expect("Unable to read file");

    for line in data.lines(){

        let parts: Vec<&str> = line.split_whitespace().collect();
        let origin = parts[0].to_string();
        let destination = parts[2].to_string();
        let cost = parts[4].parse().expect("Could not parse into number");

        nodes.entry(origin.clone()).or_insert_with(Node::new).add_neighbor(destination.clone(), cost);

        nodes.entry(destination.clone()).or_insert_with(Node::new).add_neighbor(origin.clone(), cost);
    }
    nodes
}

fn find_shortest_path(nodes: &HashMap<String, Node>) -> (Vec<String>, i32) {
    let cities: Vec<String> = nodes.keys().cloned().collect();
    let mut shortest_distance = i32::MAX;
    let mut best_route = Vec::new();

    for route in cities.iter().permutations(cities.len()){
        let mut current_distance = 0;
        let mut valid = true;

        for i in 0..route.len() -1 {
            let city = route[i];
            let next_city = route[i+1];

            if let Some(node) = nodes.get(city) {
                if let Some(&dist) = node.get_distance(next_city) {
                    current_distance += dist;
                } else {
                    valid = false;
                    break;
                }
            }
        }
        if valid && current_distance < shortest_distance {
            shortest_distance = current_distance;
            best_route = route.iter().map(|s| s.to_string()).collect();
        }
    }
    (best_route, shortest_distance)
}

fn find_longest_path(nodes: &HashMap<String, Node>) -> (Vec<String>, i32) {
    let cities: Vec<String> = nodes.keys().cloned().collect();
    let mut longest_distance = i32::MIN;
    let mut best_route = Vec::new();

    for route in cities.iter().permutations(cities.len()){
        let mut current_distance = 0;
        let mut valid = true;

        for i in 0..route.len() -1 {
            let city = route[i];
            let next_city = route[i+1];

            if let Some(node) = nodes.get(city) {
                if let Some(&dist) = node.get_distance(next_city) {
                    current_distance += dist;
                } else {
                    valid = false;
                    break;
                }
            }
        }
        if valid && current_distance > longest_distance {
            longest_distance = current_distance;
            best_route = route.iter().map(|s| s.to_string()).collect();
        }
    }
    (best_route, longest_distance)
}

fn main() {
    let nodes = parse_input("../input.txt");
    let (route, distance) = find_shortest_path(&nodes);

    println!("================= Part 1 =================");
    println!("The shortest route in the data is {} with a distance of {}", route.join(" -> "), distance);

    println!("================= Part 2 =================");
    let (route, distance) = find_longest_path(&nodes);
    println!("The longest route in the data is {} with a distance of {}", route.join(" -> "), distance);
}

File: 2015/9/go/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Node struct {
	connections map[string]int
}

func NewNode() *Node {
	return &Node{connections: make(map[string]int)}
}

func (n *Node) AddNeighbor(destination string, cost int) {
	n.connections[destination] = cost
}

func (n *Node) GetDistance(destination string) (int, bool) {
	dist, exists := n.connections[destination]
	return dist, exists
}

func ParseInput(path string) map[string]*Node {
	nodes := make(map[string]*Node)
	file, err := os.Open(path)
	if err != nil {
		panic("Unable to read file")
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parts := strings.Fields(scanner.Text())
		origin, destination := parts[0], parts[2]
		cost, _ := strconv.Atoi(parts[4])

		if _, exists := nodes[origin]; !exists {
			nodes[origin] = NewNode()
		}
		nodes[origin].AddNeighbor(destination, cost)

		if _, exists := nodes[destination]; !exists {
			nodes[destination] = NewNode()
		}
		nodes[destination].AddNeighbor(origin, cost)
	}

	return nodes
}

func Permutations(arr []string) [][]string {
	var helper func([]string, int)
	res := [][]string{}

	helper = func(arr []string, n int) {
		if n == 1 {
			tmp := make([]string, len(arr))
			copy(tmp, arr)
			res = append(res, tmp)
		} else {
			for i := 0; i < n; i++ {
				helper(arr, n-1)
				if n%2 == 1 {
					arr[0], arr[n-1] = arr[n-1], arr[0]
				} else {
					arr[i], arr[n-1] = arr[n-1], arr[i]
				}
			}
		}
	}

	helper(arr, len(arr))
	return res
}

func FindShortestPath(nodes map[string]*Node) ([]string, int) {
	cities := []string{}
	for city := range nodes {
		cities = append(cities, city)
	}

	shortestDistance := math.MaxInt32
	var bestRoute []string

	for _, route := range Permutations(cities) {
		currentDistance := 0
		valid := true

		for i := 0; i < len(route)-1; i++ {
			city, nextCity := route[i], route[i+1]
			if node, exists := nodes[city]; exists {
				if dist, exists := node.GetDistance(nextCity); exists {
					currentDistance += dist
				} else {
					valid = false
					break
				}
			}
		}

		if valid && currentDistance < shortestDistance {
			shortestDistance = currentDistance
			bestRoute = route
		}
	}

	return bestRoute, shortestDistance
}

func FindLongestPath(nodes map[string]*Node) ([]string, int) {
	cities := []string{}
	for city := range nodes {
		cities = append(cities, city)
	}

	longestDistance := math.MinInt32
	var bestRoute []string

	for _, route := range Permutations(cities) {
		currentDistance := 0
		valid := true

		for i := 0; i < len(route)-1; i++ {
			city, nextCity := route[i], route[i+1]
			if node, exists := nodes[city]; exists {
				if dist, exists := node.GetDistance(nextCity); exists {
					currentDistance += dist
				} else {
					valid = false
					break
				}
			}
		}

		if valid && currentDistance > longestDistance {
			longestDistance = currentDistance
			bestRoute = route
		}
	}

	return bestRoute, longestDistance
}

func main() {
	nodes := ParseInput("../input.txt")
	route, distance := FindShortestPath(nodes)
	fmt.Println("================= Part 1 =================")
	fmt.Printf("The shortest route in the data is %v with a distance of %d\n", strings.Join(route, " -> "), distance)

	route, distance = FindLongestPath(nodes)
	fmt.Println("================= Part 2 =================")
	fmt.Printf("The longest route in the data is %v with a distance of %d\n", strings.Join(route, " -> "), distance)
}
